### ISSUE TITLE
Linter: revive, Rule: `enforce-slice-style` - Enforces consistent usage of `make([]type, 0)`, `[]type{}`, or var `[]type` for slice initialization. Should we enable it?

### ISSUE THREAD

**Issue by zak-pawel** [influxdata/telegraf#15812]
### Description

This issue starts a discussion about enabling:
- linter: [revive](https://revive.run/) - Fast, configurable, extensible, flexible, and beautiful linter for Go. Drop-in replacement of golint. Revive provides a framework for development of custom rules, and lets you define a strict preset for enhancing your development & code review processes.
- rule: [enforce-slice-style](https://github.com/mgechev/revive/blob/master/RULES_DESCRIPTIONS.md#enforce-slice-style) - This rule enforces consistent usage of `make([]type, 0)`, `[]type{}`, or `var []type` for slice initialization. It does not affect `make([]type, non_zero_len, or_non_zero_cap)` constructions as well as `[]type{v1}`. Nil slices are always permitted.

### Configuration:

(string) Specifies the enforced style for slice initialization. The options are:
- "any": No enforcement (default).
- "make": Enforces the usage of `make([]type, 0)`.
- "literal": Enforces the usage of `[]type{}`.
- "nil": Enforces the usage of `var []type`.

#### Example configuration:

1:
```
      - name: enforce-slice-style
        arguments: ["make"]
```

2:
```
      - name: enforce-slice-style
        arguments: ["literal"]
```

3:
```
      - name: enforce-slice-style
        arguments: ["nil"]
```

### Expected output

Decision about enabling or not enabling this rule.

### Findings

For this rule, with configuration 1, `607` findings (of type: `enforce-slice-style: use make([]type) instead of []type{} (or declare nil slice)`) were found in the current codebase.

For this rule, with configuration 2, `169` findings (of type: `enforce-slice-style: use []type{} instead of make([]type, 0) (or declare nil slice)`) were found in the current codebase.

For this rule, with configuration 3, `776` findings (of type: `enforce-slice-style: use nil slice declaration (e.g. var args []type) instead of []type{}` / `enforce-slice-style: use nil slice declaration (e.g. var args []type) instead of make([]type, 0)`) were found in the current codebase.

**Comment by srebhan** [comment]
I love the `literal` style (similar to `map[abc]xyz{}` style) but it must make sense in the code... Inclined to enable this linter if we do the same for maps (#15810)...

**Comment by zak-pawel** [comment]
And I'm a fan of using `make([]type, 0)` ;)
I wonder if it makes sense to enable this rule, just to force the developer to use a version that they may not necessarily like. 
After all, `make` and `literal` versions work in exactly the same way.

**Comment by srebhan** [comment]
I think I also misunderstood this one. For me `var x []string` is the easier alternative if (and only if) the length cannot be known in advance... I prefer this because that's what you get with structs like
```golang
type xyz struct {
  name string
  properties []string
}
```
where I don't want to see an explicit init of the slice...

**Comment by zak-pawel** [comment]
@srebhan This rule enforces the initialization of **only empty slices** in a chosen way.

So:
`a := make([]string, 0)` or `a := []string{}`

Third option (`var a []string`) is not equivalent to the previous two options, because it is `nil slice`.

**Comment by srebhan** [comment]
Well but you can append to a `nil` slice without any issue (contrary to a map)... So I don't see a benefit of `a := make([]string, 0)` vs `var a []string`...

**Comment by zak-pawel** [comment]
Yes, you can `append` to all three versions. 
Each of these instructions for each of these versions will return `true`:
- `len(a) == 0`
- `cap(a) == 0`

But (obviously) only on `nil` slice you will get `true` here: `a == nil`.
And many packages do compare slices to `nil` and may act differently based on that (e.g. [encoding/json](https://pkg.go.dev/encoding/json) and [fmt](https://pkg.go.dev/fmt) packages).

Also, converting a `non-nil` slice to an array pointer will result in a `non-nil` pointer, converting a `nil` slice to an array pointer will result in a `nil` pointer (https://go.dev/ref/spec#Conversions_from_slice_to_array_or_array_pointer).

In summary, `literal` and `make` do exactly the same thing and it doesn't matter which one we use if we want to create a `non-nil` slice. 
But changing all declarations to a `nil` slice can have many (sometimes hidden) consequences.

**Comment by srebhan** [comment]
I see (and was not aware of the fact :-/). As long as declaring a `nil` slice is still possible I'm fine enforcing the `make` variant.
